src/GraphQL/Validation/Rules/NoUnusedFragments.cs

```diff
@@ -21,33 +21,41 @@ public class NoUnusedFragments : IValidationRule
 
         /// <inheritdoc/>
         /// <exception cref="NoUnusedFragmentsError"/>
-        public Task<INodeVisitor> ValidateAsync(ValidationContext context)
-        {
-            var operationDefs = new List<Operation>();
-            var fragmentDefs = new List<FragmentDefinition>();
+        public Task<INodeVisitor> ValidateAsync(ValidationContext context) => _nodeVisitor;
 
-            return new EnterLeaveListener(_ =>
+        private static readonly Task<INodeVisitor> _nodeVisitor = new NodeVisitors(
+            new MatchingNodeVisitor<Operation>((node, context) =>
             {
-                _.Match<Operation>(node => operationDefs.Add(node));
-                _.Match<FragmentDefinition>(node => fragmentDefs.Add(node));
-                _.Match<Document>(leave: document =>
+                (context.TypeInfo.NoUnusedFragments_OperationDefs ??= new List<Operation>(1)).Add(node);
+                context.TypeInfo.NoUnusedFragments_FragmentDefs?.Clear();
```

Comment:
?

I do not see the corresponding line in the previous version.

Reply:
yeah weird.  it looks like code from another rule.

Reply:
```suggestion
```

Reply:
I guess that's what happens when I refactor something three different times across 20 different files each time, all in one PR!

Reply:
It will be hard to find such errors later.